File: aoc-py/solutions/day11.py
"""
Day 11: Printing Department

https://adventofcode.com/2025/day/11
"""
__all__ = ('Day11',)

from typing import ClassVar
from functools import cache

from ..solution import Solution

class Day11(Solution):
    NAME: ClassVar[str] = 'Printing Department'
    # we must store the graph as an attribute
    # so that the DFS functions can access it without needing to pass it around
    # this also allows us to cache the arguments (since graphs which are dictionaries are not hashable)
    graph: dict[str, list[str]]

    def _parse_input(self, inp: str) -> dict[str, list[str]]:
        return {
            (parts := line.split(':'))[0]: parts[1].split()
            for line in inp.splitlines()
        }

    @cache
    def dfs_1(self, node: str) -> int:
        # base case: reached end of tree: the 'out' node, return 1 as we have completed 1 way to reach here
        if node == 'out':
            return 1
        # recursively DFS all paths to 'out' nodes
        # and sum the number of ways to them (AKA count the number of times the base-case is reached, summing up the 1s each time)
        return sum(self.dfs_1(child) for child in self.graph.get(node, []))

    @cache
    def dfs_2(self, node: str, dac: bool = False, fft: bool = False) -> int:
        # similar to `dfs` used in part 1, but now we need to ensure the nodes 'dac' and 'fft' have been visited
        # this is indicated through the booleans `dac` and `fft`
        match node:
            # return 1: valid path as we reached 'out' having visited both 'dac' and 'fft'
            case 'out' if dac and fft:
                return 1
            # return 0: this is not a valid path as we reached 'out' without visiting both 'dac' and 'fft'
            # but we still need to return to finish DFS
            case 'out':
                return 0
            # if we reach 'dac' or 'fft', set the corresponding boolean to True
            # and we continue the DFS with the variable set to True for the rest of the path
            case 'dac':
                dac = True
            case 'fft':
                fft = True

        return sum(self.dfs_2(child, dac, fft) for child in self.graph.get(node, []))

    def part_one(self, inp: str) -> int:
        self.graph = self._parse_input(inp)
        return self.dfs_1('you')

    def part_two(self, inp: str) -> int:
        self.graph = self._parse_input(inp)
        return self.dfs_2('svr')

    def run(self, inp: str) -> None:
        print('Part 1:', p1 := self.part_one(inp))
        print('Part 2:', p2 := self.part_two(inp))

        assert p1 == 758
        assert p2 == 490695961032000

File: src/bin/day11.rs
//! Day 11: Reactor
//!
//! <https://adventofcode.com/2025/day/11>
use std::collections::HashMap;
use aoc_2025::Solution;

pub struct Day11;

type Graph<'a> = HashMap<&'a str, Vec<&'a str>>;

impl Day11 {
    #[must_use]
    #[inline]
    fn parse_input(inp: &str) -> Graph<'_> {
        inp
            .lines()
            .map(|line| {
                let (from, to) = line.split_once(':').unwrap();

                (from, to.split_whitespace().collect())
            })
            .collect()
    }

    #[must_use]
    fn dfs_1<'a>(cache: &mut HashMap<&'a str, usize>, graph: &Graph<'a>, node: &'a str) -> usize {
        if let Some(&res) = cache.get(node) {
            return res;
        }

        if node == "out" {
            return 1;
        }

        let result = graph
            .get(node)
            .unwrap()
            .iter()
            .map(|&child| Self::dfs_1(cache, graph, child))
            .sum::<usize>();

        cache.insert(node, result);
        result
    }

    #[must_use]
    fn dfs_2<'a>(
        cache: &mut HashMap<(&'a str, bool, bool), usize>,
        graph: &Graph<'a>,
        node: &'a str,
        mut dac: bool,
        mut fft: bool,
    ) -> usize {
        let args = (node, dac, fft);

        if let Some(&res) = cache.get(&args) {
            return res;
        }

        match node {
            "out" if dac && fft => return 1,
            "out" => return 0,
            "dac" => dac = true,
            "fft" => fft = true,
            _ => (),
        }

        let result = graph
            .get(node)
            .unwrap()
            .iter()
            .map(|&child| Self::dfs_2(cache, graph, child, dac, fft))
            .sum::<usize>();

        cache.insert(args, result);
        result
    }
}

impl Solution for Day11 {
    const NAME: &'static str = "Reactor";

    fn part_one(&self, inp: &str) -> Self::OutputP1 {
        let graph = Self::parse_input(inp);
        let mut cache = HashMap::new();
        Self::dfs_1(&mut cache, &graph, "you")
    }

    fn part_two(&self, inp: &str) -> Self::OutputP2 {
        let graph = Self::parse_input(inp);
        let mut cache = HashMap::new();
        Self::dfs_2(&mut cache, &graph, "svr", false, false)
    }

    fn run(&self, inp: String) {
        let p1 = self.part_one(&inp);
        let p2 = self.part_two(&inp);

        println!("Part 1: {p1}");
        println!("Part 2: {p2}");

        assert_eq!(p1, 758);
        assert_eq!(p2, 490_695_961_032_000);
    }
}

fn main() {
    aoc_2025::run_day(11, &Day11);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test() { main(); }
}